src/Compilers/CSharp/Portable/BoundTree/BoundDecisionDag.cs

```diff
@@ -83,44 +92,45 @@ public ImmutableArray<BoundDecisionDag> TopologicallySortedNodes()
             // First, we topologically sort the nodes of the dag so that we can translate the nodes bottom-up.
             // This will avoid overflowing the compiler's runtime stack which would occur for a large switch
             // statement if we were using a recursive strategy.
-            ImmutableArray<BoundDecisionDag> sortedNodes = this.TopologicallySortedNodes();
+            ImmutableArray<BoundDecisionDagNode> sortedNodes = this.TopologicallySortedNodes;
 
             // Cache simplified/translated replacement for each translated dag node. Since we always visit
             // a node's successors before the node, the replacement should always be in the cache when we need it.
-            var replacement = PooledDictionary<BoundDecisionDag, BoundDecisionDag>.GetInstance();
+            var replacement = PooledDictionary<BoundDecisionDagNode, BoundDecisionDagNode>.GetInstance();
 
             // Loop backwards through the topologically sorted nodes to translate them, so that we always visit a node after its successors
             for (int i = sortedNodes.Length - 1; i >= 0; i--)
             {
-                BoundDecisionDag node = sortedNodes[i];
+                BoundDecisionDagNode node = sortedNodes[i];
                 Debug.Assert(!replacement.ContainsKey(node));
-                BoundDecisionDag newNode = makeReplacement(node);
+                BoundDecisionDagNode newNode = makeReplacement(node);
                 replacement.Add(node, newNode);
             }
 
-            var result = replacement[this];
+            // Return the computed replacement root node
+            var newRoot = replacement[this.RootNode];
             replacement.Free();
-            return result;
+            return this.Update(newRoot);
 
             // Make a replacement for a given node, using the precomputed replacements for its successors.
-            BoundDecisionDag makeReplacement(BoundDecisionDag dag)
+            BoundDecisionDagNode makeReplacement(BoundDecisionDagNode dag)
             {
                 switch (dag)
                 {
-                    case BoundEvaluationPoint p:
+                    case BoundEvaluationDecisionDagNode p:
```

Comment:
Are these types sealed?   (for faster typechecks)

Reply:
The bound node generator only generates `abstract` or `sealed` types.

---
In reply to: [178199580](https://github.com/dotnet/roslyn/pull/25697#discussion_r178199580) [](ancestors = 178199580)